### PR TITLE
Nimshi89 #641 new validation method

### PR DESCRIPTION
**Pull request by nimshi89**
#641 
PR opened for testing purposes.

### CODE REVIEW

**Comment by nimshi89** [comment]
I have added an `def __init_subclass__` method to the Validationtester subclass and now the SonarCloud is not complaining. Not sure why the constructor was needed and that .super() return value is new to me. But all the checks pass.

File: src/csvcubed/models/validatedmodel.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# this class will be class attributes and returns an array of errors
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:1]: Can you make sure you use a docstring in [the appropriate format](https://github.com/GSS-Cogs/csvcubed/blob/main/docs/developer.md#api-documentation) here?

File: src/csvcubed/models/validatedmodel.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from csvcubed.utils.validations import (
+    validate_str_type,
+    validate_int_type,
+    validate_list,
+)
+
+ValidationFunction = Callable[[Any, str], List[ValidateModelProperiesError]]
+
+
+class ValidatedModel(DataClassBase):
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:18]: Can you add a docstring here explaining the purpose of the `ValidatedModel` and can you make it an abstract base class too so it's not possible to directly instantiate it?

File: src/csvcubed/models/validatedmodel.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    def validate(self) -> List[ValidateModelProperiesError]:
+        validation_errors: List[ValidateModelProperiesError] = []
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:20]: Can you add a doc string here explaining what the `validate` function will do?

File: src/csvcubed/models/validatedmodel.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        for (property_name, validation_function) in self._get_validations().items():
+            print(f"Validating '{property_name}'")
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:22]: Can you convert this to a `_logger.debug` call instead of print? And make sure to use the `%s` syntax for string building.

File: src/csvcubed/models/validatedmodel.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            property_value = getattr(self, property_name)
+            validation_errors += validation_function(property_value, property_name)
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:24]: Can you put some debug logging after calling `validation_function` to make it clear which property's validation function generated which errors?

File: src/csvcubed/models/validatedmodel.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+        return validation_errors
+
+    @abstractmethod
+    def _get_validations(self) -> Dict[str, ValidationFunction]:
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:29]: Again, can you add a docstring comment here explaining what needs to be implemented - perhaps provide an example?

File: src/csvcubed/models/validatedmodel.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        pass
+
+
+@dataclass
+class ValidationTester(ValidatedModel):
+    string_variable: str
+
+    def __init_subclass__(cls) -> None:
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:37]: I don't see why we'd need this, perhaps we can discuss it?

File: src/csvcubed/models/validatedmodel.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# this class will be class attributes and returns an array of errors
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List
+
+from csvcubedmodels.dataclassbase import DataClassBase
+from csvcubed.models.validationerror import ValidateModelProperiesError
+from csvcubed.utils.validations import (
+    validate_str_type,
```
Comment (robons) [inline src/csvcubed/models/validatedmodel.py:10]: Can you only import the functions we definitely need to use? I know the int and list ones aren't necessary here.

File: src/csvcubed/models/validationerror.py [hunk 9]
```diff
@@ -60,3 +60,11 @@ class PydanticThrowableSpecificValidationError(
...
 
     pass
+
+@dataclass
+class ValidateModelProperiesError(ValidationError):
+
+    """
+    This error will be returned with the name of the class attribute.
+    """
+    property_name: str
```
Comment (robons) [inline src/csvcubed/models/validationerror.py:70]: Can you create a ticket to expand the information we hold here? I think it'd be useful to include a path to the property instead of just a single property_name.

File: src/csvcubed/models/validationerror.py [hunk 10]
```diff
@@ -60,3 +60,11 @@ class PydanticThrowableSpecificValidationError(
     """
 
     pass
+
+@dataclass
+class ValidateModelProperiesError(ValidationError):
+
+    """
+    This error will be returned with the name of the class attribute.
```
Comment (robons) [inline src/csvcubed/models/validationerror.py:68]: Can you explain a bit better the situation/context in which this error will be generated?

File: src/csvcubed/utils/validations.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+from csvcubed.utils.uri import looks_like_uri
+
+
+T = TypeVar("T")
+
+
+def validate_list(
+    validate_list_item: Callable[[T], List[ValidateModelProperiesError]],
+) -> Callable[[List[T], str], List[ValidateModelProperiesError]]:
```
Comment (robons) [inline src/csvcubed/utils/validations.py:15]: Can you add docstring comments to all of these functions to explain what they do?

File: tests/pref_improvement/pref_check.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+import inspect
```
Comment (robons) [inline tests/pref_improvement/pref_check.py:1]: Again, can you put a docstring at the top explaining what this module is for?

File: tests/pref_improvement/pref_check.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import sys
+from typing import Iterable
+from dataclasses import is_dataclass, fields
+from pathlib import Path
+
+import csvcubed
+from csvcubed.definitions import APP_ROOT_DIR_PATH
+from csvcubed.models.validatedmodel import ValidatedModel
+
+# scan throught each file and check if the members are classes
```
Comment (robons) [inline tests/pref_improvement/pref_check.py:12]: These comments should be docstring comments below the function's definition. This way the information will show up in your IDE's intellisense.

File: tests/pref_improvement/validator.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    @abstractmethod
+    def _get_validations(self) -> Dict[str, Callable[[Any], List[ValidationErrors]]]:
+        pass
+
+    @abstractmethod
+    def ensure_str(self):
+        pass
+    @abstractmethod
+    def ensure_positive_integer_or_none(self):
+        pass
```
Comment (robons) [inline tests/pref_improvement/validator.py:24]: Do we need this file?

File: tests/test-cases/test_class_collector/test_file.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
+
+from dataclasses import dataclass
+
+
+class Myclass:
```
Comment (robons) [inline tests/test-cases/test_class_collector/test_file.py:5]: Is this file necessary?

File: tests/unit/models/test_model.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    def _get_validations(self) -> Dict[str, ValidationFunction]:
+        return {
+            "str_test_variable": validate_str_type,
+            "int_test_variable": validate_int_type,
+            "list_test_variable": validate_list(validate_str_type),
+            "test_uri": validate_optional(validate_uri),
+        }
+
+
+def test_validate_int_type_incorrect():
```
Comment (robons) [inline tests/unit/models/test_model.py:35]: Can you provide doc string comments explaining each of these tests?

File: tests/unit/pref_tests/test_pref_check.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    check_for_dataclass,
+    _map_the_thing,
+    _get_all_classes,
+    acess_all_folders_return_all_files,
+)
+
+from pathlib import Path
+
+# test to ensure it returns the corrects name
+def test_map_that_thing():
```
Comment (robons) [inline tests/unit/pref_tests/test_pref_check.py:16]: Please rename that function.

**Review by robons — CHANGES_REQUESTED** [review]
Good work, but it needs a bit of tidying up before it's ready to be merged. Let's talk aboubt anything you're unsure about later.

File: .github/workflows/reusable-test.yaml [hunk 18]
```diff
@@ -54,10 +54,14 @@ jobs:
         run: poetry install
 
       - name: Run pyright
-        run: poetry run pyright . --lib
+        run: poetry run pyright src/ --lib
+      
+      - name: Run validation
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply .github/workflows/reusable-test.yaml:59]: `- name: Run validation for all class attributes and variables that has a dataclass decorator, before build`
would this be better or too much?

File: pyproject.toml [hunk 19]
```diff
@@ -40,7 +40,6 @@ requires = ["poetry-core>=1.0.0"]
 [tool.pyright]
 exclude = [
   ".venv",
-  "**/tests",
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply pyproject.toml:43]: oh yeah so pytest will ignore the test files and not complain about them

File: src/csvcubed/models/oriigami.py [hunk 20]
```diff
@@ -0,0 +1,5 @@
+from dataclasses import dataclass
+
+@dataclass
+class CheeseBurger:
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply src/csvcubed/models/oriigami.py:4]: done :D

File: src/csvcubed/models/validatedmodel.py (hunk 2 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:18]: done

File: src/csvcubed/models/validatedmodel.py (hunk 3 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:20]: done

File: src/csvcubed/models/validatedmodel.py (hunk 5 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:24]: done

File: src/csvcubed/models/validatedmodel.py (hunk 6 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:29]: done

File: src/csvcubed/models/validatedmodel.py (hunk 7 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:37]: yeah when you are free just call me. I did struggle with that

File: src/csvcubed/models/validatedmodel.py (hunk 8 above)
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:10]: done

File: src/csvcubed/models/validationerror.py (hunk 9 above)
Comment (nimshi89) [reply src/csvcubed/models/validationerror.py:70]: Isn't that provided in the script that utilises these functions? `pref_check.py`

` print(
                f"In {j.__name__} ({inspect.getfile(j)}) you have not validated {', '.join(not_validated)}"
            )
            print(format(field_names))  # This returns empty`

this will print out the variable names that need validation and the file path.  but i can of course make a ticket if it's still required.

File: src/csvcubed/utils/validations.py (hunk 11 above)
Comment (nimshi89) [reply src/csvcubed/utils/validations.py:15]: done

File: tests/pref_improvement/pref_check.py (hunk 12 above)
Comment (nimshi89) [reply tests/pref_improvement/pref_check.py:1]: done

File: src/csvcubed/models/validationerror.py (hunk 10 above)
Comment (nimshi89) [reply src/csvcubed/models/validationerror.py:68]: done

File: tests/pref_improvement/validator.py (hunk 14 above)
Comment (nimshi89) [reply tests/pref_improvement/validator.py:24]: nope. sorted

File: tests/test-cases/test_class_collector/test_file.py (hunk 15 above)
Comment (nimshi89) [reply tests/test-cases/test_class_collector/test_file.py:5]: Yes this file is used for unit tests

File: src/csvcubed/models/cube/qb/components/codelist.py [hunk 21]
```diff
@@ -179,3 +180,16 @@ class CompositeQbCodeList(NewQbCodeList[DuplicatedQbConcept]):
...
+    item: QbCodeList, property_name: str
+) -> List[ValidateModelProperiesError]:
+    if not isinstance(item, QbCodeList):
+        return [
+            ValidateModelProperiesError(
+                f"This variable should be a QbCodeList, check the following variable:",
+                property_name,
+            )
+        ]
+    return []
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply src/csvcubed/models/cube/qb/components/codelist.py:195]: done

File: .github/workflows/reusable-test.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (nimshi89) [reply .github/workflows/reusable-test.yaml]: done

File: tests/unit/pref_tests/test_pref_check.py (hunk 17 above)
Comment (nimshi89) [reply tests/unit/pref_tests/test_pref_check.py:16]: done

File: src/csvcubed/models/validationerror.py (hunk 9 above)
Comment (nimshi89) [reply src/csvcubed/models/validationerror.py:72]: for example what data should be stored in that class.
```suggestion
    property_path: List[str]
```

e.g. if we have the following:

```json
{
  "chicken": {
    "burgers": [
      {
        "should_be_a_string": 4
      },
      {
        "should_be_a_string": "woof"
      }
    ]
  }
}
```

The property path for the offending validation error should be ["chicken", "burgers", "0", "should_be_a_string"]

File: src/csvcubed/utils/validations.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+from typing import List, TypeVar, Type, Callable
+
+from csvcubed.utils.uri import looks_like_uri
+
+
+T = TypeVar("T")
+
+"""
+This function will valide if the argument provided is in fact a list and,
+in a loop will check each member of the list and return an error if the item in the list is not the correct type.
```
Comment (robons) [inline src/csvcubed/utils/validations.py:14]: Should say something like: ` and returns any errors returned by the item validation function`

File: tests/pref_improvement/pref_check.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def main():
+    # listing all the classes from each file in the directory and subdirectory given.
+    all_classes = _get_all_classes(APP_ROOT_DIR_PATH)
+
+    # list all dataclasses(the classes that have the datacall decorator)
+    all_data_classes = check_for_dataclass(all_classes)
+
+    validated_model_classes = []
```
Comment (robons) [inline tests/pref_improvement/pref_check.py:90]: Just an FYI that since you're building up a list here, you lose the benefits of everything being iterable up to this point - you have to wait for everything to complete now before you can go on to the next step.

File: tests/pref_improvement/pref_check.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+        field_names = {f.name for f in fields(validated_model)}
+
+        if field_names_validated != field_names:
+
+            not_validated = field_names - field_names_validated
+            print(
+                f"In {validated_model_class.__name__} ({inspect.getfile(validated_model_class)}) you have not validated {', '.join(not_validated)}"
+            )
+            print(format(field_names))  # This returns empty
+            exit(1)
```
Comment (robons) [inline tests/pref_improvement/pref_check.py:107]: Just a reminder that the position of this `exit(1)` means that we only ever print one error to the console. It might be nicer to continue going and then ensure the app does the `exit(1)` if we had to print any errors to the console. 

I often find it helpful that a test tells me about all of the errors rather than just the first one.

**Review by robons — COMMENTED** [review]
Nearly there.

See [here](https://www.programiz.com/python-programming/docstrings) for some examples of how docstrings work in python.

File: .github/workflows/reusable-test.yaml [hunk 25]
```diff
@@ -54,10 +54,14 @@ jobs:
         run: poetry install
 
       - name: Run pyright
-        run: poetry run pyright . --lib
+        run: poetry run pyright src/ --lib
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply .github/workflows/reusable-test.yaml:57]: done

File: src/csvcubed/models/validatedmodel.py [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+ValidationFunction = Callable[[Any, str], List[ValidateModelProperiesError]]
+
+"""This abrstract class that will act as a parent class for class attribute validations.
+The _get_validations function will return a dictionary containing the names of the variables and corespondinv validatin methods.
+"""
+
+
+class ValidatedModel(DataClassBase):
+    """The validate function will go through each attribue and the corresponding validation function and collect the validation errors(if there is any) and return the variable names and the error messages"""
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply src/csvcubed/models/validatedmodel.py:22]: done

File: tests/unit/models/test_model.py [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+    validate_int_type,
+    validate_str_type,
+    validate_list,
+    validate_uri,
+    validate_optional,
+)
+from csvcubed.models.validatedmodel import ValidatedModel
+from csvcubed.models.validationerror import ValidateModelProperiesError
+
+ValidationFunction = Callable[[Any, str], List[ValidateModelProperiesError]]
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply tests/unit/models/test_model.py:16]: sorted

File: tests/unit/readers/cubeconfig/v1_0/test_build_friendly_error_msgs.py [hunk 28]
```diff
@@ -458,31 +460,31 @@ def test_val_errors_undefined_unit_uri():
...
-    _assert_in_log(
-        "csvcubed.cli.build - ERROR - Validation Error: A URI collision has been detected in an attribute column."
-    )
-    _assert_in_log(
-        "The values 'Software Sales', 'software-sales' map to the same URI-safe identifier 'software-sales'"
-    )
-    _assert_in_log(
-        "csvcubed.cli.build - ERROR - More information: https://purl.org/csv-cubed/err/conflict-uri"
-    )
+# def test_val_errors_uri_conflict():
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply tests/unit/readers/cubeconfig/v1_0/test_build_friendly_error_msgs.py:463]: done

File: tests/pref_improvement/pref_check.py (hunk 23 above)
Comment (nimshi89) [reply tests/pref_improvement/pref_check.py:90]: done

File: tests/pref_improvement/pref_check.py (hunk 24 above)
Comment (nimshi89) [reply tests/pref_improvement/pref_check.py]: Okay I implemented a solution and committed it. Hope it will be as should

File: src/csvcubed/utils/validations.py [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+
+from typing import List, TypeVar, Type, Callable
+
+from csvcubed.utils.uri import looks_like_uri
+
+
+T = TypeVar("T")
+
+
+"""
```
Comment (robons) [inline src/csvcubed/utils/validations.py:13]: Still in the wrong place.
